spring-kafka/src/main/java/org/springframework/kafka/annotation/BackOffFactory.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2018-present the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.annotation;
+
+import java.time.Duration;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+import org.jspecify.annotations.Nullable;
+
+import org.springframework.core.retry.RetryPolicy;
+import org.springframework.format.annotation.DurationFormat;
+import org.springframework.format.datetime.standard.DurationFormatterUtils;
+import org.springframework.util.Assert;
+import org.springframework.util.StringUtils;
+import org.springframework.util.StringValueResolver;
+import org.springframework.util.backoff.FixedBackOff;
+
+/**
+ * Create a {@link org.springframework.util.backoff.BackOff} from the state of a
+ * {@link BackOff @BackOff} annotation.
+ *
+ * @author Stephane Nicoll
+ */
+final class BackOffFactory {
+
+	private static final long DEFAULT_DELAY = 1000;
+
+	private final @Nullable StringValueResolver embeddedValueResolver;
+
+	BackOffFactory(@Nullable StringValueResolver embeddedValueResolver) {
+		this.embeddedValueResolver = embeddedValueResolver;
+	}
+
+	/**
+	 * Create a {@link org.springframework.util.backoff.BackOff} instance based on the
+	 * state of the given {@link BackOff @Backff}. The returned backoff instance has
+	 * unlimited number of attempts as these are controlled externally.
+	 *
+	 * @param annotation the annotation to source the parameters from
+	 * @return a {@link org.springframework.util.backoff.BackOff}
+	 */
+	public org.springframework.util.backoff.BackOff createFromAnnotation(BackOff annotation) { // NOSONAR
```

Comment:
The `public` is redundant here.
And I agree that no reason to make this class as `public`.

Reply:
Good catch. There is a rule in Spring Boot if you want to enforce it. I think it might be ArchUnit. LMK if you're interested.